kani-compiler/src/kani_middle/transform/check_uninit/uninit_visitor.rs

```diff
@@ -0,0 +1,712 @@
+    /// When we instrument an instruction, we partition the basic block, and the instruction that
+    /// may trigger UB becomes the first instruction of the basic block, which we need to skip
+    /// later.
+    skip_next: bool,
+    /// The instruction being visited at a given point.
+    current: SourceInstruction,
+    /// The target instruction that should be verified.
+    pub target: Option<InitRelevantInstruction>,
+    /// The basic block being visited.
+    bb: BasicBlockIdx,
+}
+
+impl<'a> CheckUninitVisitor<'a> {
+    pub fn find_next(
+        body: &'a MutableBody,
+        bb: BasicBlockIdx,
+        skip_first: bool,
+    ) -> Option<InitRelevantInstruction> {
+        let mut visitor = CheckUninitVisitor {
+            locals: body.locals(),
+            skip_next: skip_first,
+            current: SourceInstruction::Statement { idx: 0, bb },
+            target: None,
+            bb,
+        };
+        visitor.visit_basic_block(&body.blocks()[bb]);
+        visitor.target
+    }
+
+    fn push_target(&mut self, source_op: MemoryInitOp) {
+        let target = self.target.get_or_insert_with(|| InitRelevantInstruction {
+            source: self.current,
+            after_instruction: vec![],
+            before_instruction: vec![],
+        });
+        target.push_operation(source_op);
+    }
+}
+
+impl<'a> MirVisitor for CheckUninitVisitor<'a> {
+    fn visit_statement(&mut self, stmt: &Statement, location: Location) {
+        if self.skip_next {
+            self.skip_next = false;
+        } else if self.target.is_none() {
+            // Leave it as an exhaustive match to be notified when a new kind is added.
+            match &stmt.kind {
+                StatementKind::Intrinsic(NonDivergingIntrinsic::CopyNonOverlapping(copy)) => {
+                    self.super_statement(stmt, location);
+                    // Source is a *const T and it must be initialized.
+                    self.push_target(MemoryInitOp::Check {
+                        operand: copy.src.clone(),
+                        count: copy.count.clone(),
+                    });
+                    // Destimation is a *mut T so it gets initialized.
+                    self.push_target(MemoryInitOp::Set {
+                        operand: copy.dst.clone(),
+                        count: copy.count.clone(),
+                        value: true,
+                        position: InsertPosition::After,
+                    });
+                }
+                StatementKind::Assign(place, rvalue) => {
+                    // First check rvalue.
+                    self.visit_rvalue(rvalue, location);
+                    // Check whether we are assigning into a dereference (*ptr = _).
+                    if let Some(place_without_deref) = try_remove_topmost_deref(place) {
+                        // First, check that we are not dereferencing extra pointers along the way
```

Comment:
BTW, you could reuse the `visit_place` code. But not a blocker.

Reply:
You're right -- I'll fix it in the follow-up PR